File: compiler/DictNode.clj
(
    (Import "CompilerGeneric.clj")

    (Enum DictNodeType:int
        (
            NoType
            Double
            Integer
            String
            Boolean
            Array
            Object
            Null
        )
    )
    ( CreateClass DictNode:void 
        (

            (def is_property:boolean false)
            (def is_property_value:boolean false)

            (def vref:string)
            (def value_type:int DictNodeType.NoType)

            (def double_value:double)
            (def string_value:string)
            (def boolean_value:boolean)
            (def object_value:DictNode)

            (def children:[DictNode])
            (def objects:[string:DictNode])
            (def keys:[string])

            (PublicMethod EncodeString:string (orig_str:string)
                (
                    (def encoded_str:string "")
                    (def str_length:int (strlen orig_str))
                    (def ii:int 0)

                    ; a bit slow algorithm, could be improved much faster by copying
                    ; longer slices 
                    (while (< ii str_length)
                        (
                            (def cc:char (charAt orig_str ii))

                            (switch cc
                                (case 8 
                                    ( = encoded_str (+ encoded_str (strfromcode 92 ) (strfromcode 98 ) ) ) 
                                )    
                                (case 9 
                                    ( = encoded_str (+ encoded_str (strfromcode 92 ) (strfromcode 116 ) ) ) 
                                )    
                                (case 10 
                                    ( = encoded_str (+ encoded_str (strfromcode 92 ) (strfromcode 110 ) ) ) 
                                )    
                                (case 12 
                                    ( = encoded_str (+ encoded_str (strfromcode 92 ) (strfromcode 102 ) ) ) 
                                )    
                                (case 13 
                                    ( = encoded_str (+ encoded_str (strfromcode 92 ) (strfromcode 114 ) ) ) 
                                )    
                                (case 34 
                                    ( = encoded_str (+ encoded_str (strfromcode 92 ) (strfromcode 34 ) ) ) 
                                )  
                                (case 92 
                                    ( = encoded_str (+ encoded_str (strfromcode 92 ) (strfromcode 92 ) ) ) 
                                )
                                (case 47 
                                    ( = encoded_str (+ encoded_str (strfromcode 92 ) (strfromcode 47 ) ) ) 
                                )                                    
                             
                                (default
                                    ( = encoded_str (+ encoded_str (strfromcode cc) ) ) 
                                )
                            )   
                            (= ii (+ 1 ii))
                        )
                    )
                    (return encoded_str)   
                )
            )            

            ( PublicMethod createEmptyObject:DictNode () (
                (def v:DictNode (new DictNode ()))
                (= v.value_type DictNodeType.Object)
                (return v)
            ))

            ( PublicMethod addObject:DictNode (key:string) (
                (if (== value_type DictNodeType.Object)
                    (
                        (def p:DictNode (new DictNode ()))                       
                        (def v:DictNode (new DictNode ()))
                        (= p.value_type DictNodeType.Object)
                        (= p.vref key)
                        (= p.is_property true)
                        (= v.value_type DictNodeType.Object)
                        (= v.vref key)
                        (= v.is_property_value true)
                        (= p.object_value v)
                        (push keys key) 
                        (set objects key p)       

                        (return v)  
                    )
                )
            ))
            

            ( PublicMethod addString:void (key:string value:string) (
                (if (== value_type DictNodeType.Object)
                    (
                    
                        (def v:DictNode (new DictNode ()))
                        (= v.string_value value)
                        (= v.value_type DictNodeType.String)
                        (= v.vref key)
                        (= v.is_property true)
                        (push keys key) 
                        (set objects key v)         
                    )
                )
            ))
            

            ( PublicMethod getDoubleAt:double (index:int) (
            
                (if (< index (array_length children))
                    (
                        (def k:DictNode (itemAt children index))
                        (return k.double_value)
                    )
                )
                (return 0)
            ))
            ;
            ( PublicMethod getDouble:double (key:string) (
                (if (has objects key)
                    (
                        (def k:DictNode (get objects key))
                        (return k.double_value)
                    )
                )
                (return 0)
            ))

            ( PublicMethod getStringAt:string (index:int) (
            
                (if (< index (array_length children))
                    (
                        (def k:DictNode (itemAt children index))
                        (return k.string_value)
                    )
                )
                (return "")
            ))
            ;
            ( PublicMethod getString:string (key:string) (
                (if (has objects key)
                    (
                        (def k:DictNode (get objects key))                     
                        (return k.string_value)
                    )
                )
                (return "")
            ))

            ( PublicMethod getArray:DictNode (key:string) (
                (if (has objects key)
                    (
                        (def obj:DictNode (get objects key))
                        (if obj.is_property
                            (return obj.object_value)
                        )
                    )
                )
                (return  (new DictNode () ))
            ))  

            ( PublicMethod getArrayAt:DictNode (index:int) (
                (if (< index (array_length children))
                    (
                        (def k:DictNode (itemAt children index))
                        (return k)
                    )
                )
                (return  (new DictNode ()))
            ))    

            ( PublicMethod getObject:DictNode (key:string) (
                (if (has objects key)
                    (
                        (def obj:DictNode (get objects key))
                        (if obj.is_property
                            (return obj.object_value)
                        )
                    )
                )
                (return  (new DictNode ()))
            ))  

            ( PublicMethod getObjectAt:DictNode (index:int) (

                (if (< index (array_length children))
                    (
                        (def k:DictNode (itemAt children index))
                        (return k)
                    )
                )
                (return  (new DictNode ( )))
            ))                                  
            
            ( PublicMethod walk:void () (

                (if is_property
                    (
                        (if (== value_type DictNodeType.Null)
                            (print (+ vref " : null"))
                        )                
                        (if (== value_type DictNodeType.Double)
                            (print (+ vref " : " double_value))
                        )
                        (if (== value_type DictNodeType.String)
                            (print (+ vref " : " string_value))
                        )
                        (if (== value_type DictNodeType.Boolean)
                            (if boolean_value
                                (print (+ vref " : " "true"))
                                (print (+ vref " : " "false"))
                            )
                        )
                    )
                    (
                        (if (== value_type DictNodeType.Null)
                            (print "null")
                        )                
                        (if (== value_type DictNodeType.Double)
                            (print double_value)
                        )
                        (if (== value_type DictNodeType.String)
                            (print string_value)
                        )
                        (if (== value_type DictNodeType.Boolean)
                            (if boolean_value
                                (print "true")
                                (print "false")
                            )
                        )
                    )                    
                )


                (if (== value_type DictNodeType.Array)
                    (
                        (if is_property
                            (print (+ vref " : Array("))
                            (print "Array(")
                        )
                        (print (+ "child cnt == " (array_length children)))
                        (for children item:DictNode i 
                            (
                                (call item walk ())
                                (print ",")
                            )
                        )   
                        (print ")") 
                    )
                )

                (if (== value_type DictNodeType.Object)
                    (
                        ; (print (+ "obj child cnt == " (array_length children)))
                        (if is_property
                            (
                                (print (+ vref " : "))
                                (call object_value walk ())
                            )
                            (
                                (print "Object(")
                                (for keys key:string i 
                                    (
                                        (def item:DictNode (get objects key))
                                        (call item walk ())
                                        (print ",")
                                    )
                                ) 
                                (print ")") 
                            )
                        )
                              
                        
                    )
                )                

            ))

            ( PublicMethod stringify:string () (

                (if is_property
                    (
                        (if (== value_type DictNodeType.Null)
                            (return (+ (strfromcode 34) vref (strfromcode 34) ":null"))
                        )                
                        (if (== value_type DictNodeType.Double)
                            (return (+ (strfromcode 34) vref (strfromcode 34) ":" double_value))
                        )
                        (if (== value_type DictNodeType.String)
                            (return (+ (strfromcode 34) vref (strfromcode 34) ":" (strfromcode 34) ( EncodeString string_value ) (strfromcode 34) ))
                        )
                        (if (== value_type DictNodeType.Boolean)
                            (if boolean_value
                                (return (+ (strfromcode 34) vref (strfromcode 34) ":" "true"))
                                (return (+ (strfromcode 34) vref (strfromcode 34) ":" "false"))
                            )
                        )
                    )
                    (
                        (if (== value_type DictNodeType.Null)
                            (return "null")
                        )                
                        (if (== value_type DictNodeType.Double)
                            (return ( + "" double_value) )
                        )
                        (if (== value_type DictNodeType.String)
                            (return ( + (strfromcode 34) ( EncodeString string_value ) (strfromcode 34)))
                        )
                        (if (== value_type DictNodeType.Boolean)
                            (if boolean_value
                                (return "true")
                                (return "false")
                            )
                        )
                    )                    
                )


                (if (== value_type DictNodeType.Array)
                    (
                        (def str:string "")
                        (if is_property
                            (
                                (= str (+ (strfromcode 34) vref (strfromcode 34) ":["))                            
                            )
                            (
                                (= str "[")
                            )
                        )
                        (for children item:DictNode i 
                            (
                                (if (> i 0)
                                    (= str (+ str ","))
                                )
                                (= str (+ str (call item stringify ()) ))
                            )
                        )   
                        (= str (+ str "]"))
                        (return str)
                    )
                )

                (if (== value_type DictNodeType.Object)
                    (
                        (def str:string "")

                        (if is_property
                            (
                                (return (+ (strfromcode 34) vref (strfromcode 34) ":"
                                    (call object_value stringify ())
                                ))
                                
                            )
                            (
                                (= str "{")
                                (for keys key:string i 
                                    (
                                        (if (> i 0)
                                            (= str (+ str ","))
                                        )
                                        (def item:DictNode (get objects key))
                                        (= str ( + str (call item stringify ())))
                                    )
                                ) 
                                (= str (+ str "}"))
                                (return str) 
                            )
                        )                        
                        
                    )
                )       
                (return "")         

            ))


        )
    )
    
    
)

File: versions/java/DictNode.java
import java.util.Optional;
import java.util.*;

class DictNode { 
  public boolean is_property = false;
  public boolean is_property_value = false;
  public String vref = "";
  public int value_type = 6;
  public double double_value = 0;
  public int int_value = 0;
  public String string_value = "";
  public boolean boolean_value = false;
  public Optional<DictNode> object_value = Optional.empty();
  public ArrayList<DictNode> children = new ArrayList<DictNode>();
  public HashMap<String,DictNode> objects = new HashMap<String,DictNode>();
  public ArrayList<String> keys = new ArrayList<String>();
  
  public static DictNode createEmptyObject() {
    final DictNode v = new DictNode();
    v.value_type = 6;
    return v;
  }
  
  
  public String EncodeString( String orig_str ) {
    String encoded_str = "";
    /** unused:  final int str_length = orig_str.length()   **/ ;
    int ii = 0;
    final byte[] buff = orig_str.getBytes();
    final int cb_len = buff.length;
    while (ii < cb_len) {
      final byte cc = buff[ii];
      switch (cc ) { 
        case 8 : 
          encoded_str = (encoded_str + ((new String( Character.toChars(92))))) + ((new String( Character.toChars(98))));
          break;
        case 9 : 
          encoded_str = (encoded_str + ((new String( Character.toChars(92))))) + ((new String( Character.toChars(116))));
          break;
        case 10 : 
          encoded_str = (encoded_str + ((new String( Character.toChars(92))))) + ((new String( Character.toChars(110))));
          break;
        case 12 : 
          encoded_str = (encoded_str + ((new String( Character.toChars(92))))) + ((new String( Character.toChars(102))));
          break;
        case 13 : 
          encoded_str = (encoded_str + ((new String( Character.toChars(92))))) + ((new String( Character.toChars(114))));
          break;
        case 34 : 
          encoded_str = (encoded_str + ((new String( Character.toChars(92))))) + "\"";
          break;
        case 92 : 
          encoded_str = (encoded_str + ((new String( Character.toChars(92))))) + ((new String( Character.toChars(92))));
          break;
        case 47 : 
          encoded_str = (encoded_str + ((new String( Character.toChars(92))))) + ((new String( Character.toChars(47))));
          break;
        default: 
          encoded_str = encoded_str + ((new String( new char[] {cc })));
          break;
      }
      ii = 1 + ii;
    }
    return encoded_str;
  }
  
  public void addString( String key , String value ) {
    if ( value_type == 6 ) {
      final DictNode v = new DictNode();
      v.string_value = value;
      v.value_type = 3;
      v.vref = key;
      v.is_property = true;
      keys.add(key);
      objects.put(key, v);
    }
  }
  
  public void addDouble( String key , double value ) {
    if ( value_type == 6 ) {
      final DictNode v = new DictNode();
      v.double_value = value;
      v.value_type = 1;
      v.vref = key;
      v.is_property = true;
      keys.add(key);
      objects.put(key, v);
    }
  }
  
  public void addInt( String key , int value ) {
    if ( value_type == 6 ) {
      final DictNode v = new DictNode();
      v.int_value = value;
      v.value_type = 2;
      v.vref = key;
      v.is_property = true;
      keys.add(key);
      objects.put(key, v);
    }
  }
  
  public void addBoolean( String key , boolean value ) {
    if ( value_type == 6 ) {
      final DictNode v = new DictNode();
      v.boolean_value = value;
      v.value_type = 4;
      v.vref = key;
      v.is_property = true;
      keys.add(key);
      objects.put(key, v);
    }
  }
  
  public Optional<DictNode> addObject( String key ) {
    Optional<DictNode> v = Optional.empty();
    if ( value_type == 6 ) {
      final DictNode p = new DictNode();
      v = Optional.of(new DictNode());
      p.value_type = 6;
      p.vref = key;
      p.is_property = true;
      v.get().value_type = 6;
      v.get().vref = key;
      v.get().is_property_value = true;
      p.object_value = v;
      keys.add(key);
      objects.put(key, p);
      return Optional.ofNullable((v.isPresent() ? (DictNode)v.get() : null ) );
    }
    return Optional.ofNullable((v.isPresent() ? (DictNode)v.get() : null ) );
  }
  
  public void setObject( String key , DictNode value ) {
    if ( value_type == 6 ) {
      final DictNode p = new DictNode();
      p.value_type = 6;
      p.vref = key;
      p.is_property = true;
      value.is_property_value = true;
      value.vref = key;
      p.object_value = Optional.of(value);
      keys.add(key);
      objects.put(key, p);
    }
  }
  
  public Optional<DictNode> addArray( String key ) {
    Optional<DictNode> v = Optional.empty();
    if ( value_type == 6 ) {
      v = Optional.of(new DictNode());
      v.get().value_type = 5;
      v.get().vref = key;
      v.get().is_property = true;
      keys.add(key);
      objects.put(key, v.get());
      return Optional.ofNullable((v.isPresent() ? (DictNode)v.get() : null ) );
    }
    return Optional.ofNullable((v.isPresent() ? (DictNode)v.get() : null ) );
  }
  
  public void push( DictNode obj ) {
    if ( value_type == 5 ) {
      children.add(obj);
    }
  }
  
  public double getDoubleAt( int index ) {
    if ( index < (children.size()) ) {
      final DictNode k = children.get(index);
      return k.double_value;
    }
    return 0;
  }
  
  public String getStringAt( int index ) {
    if ( index < (children.size()) ) {
      final DictNode k = children.get(index);
      return k.string_value;
    }
    return "";
  }
  
  public int getIntAt( int index ) {
    if ( index < (children.size()) ) {
      final DictNode k = children.get(index);
      return k.int_value;
    }
    return 0;
  }
  
  public boolean getBooleanAt( int index ) {
    if ( index < (children.size()) ) {
      final DictNode k = children.get(index);
      return k.boolean_value;
    }
    return false;
  }
  
  public Optional<String> getString( String key ) {
    Optional<String> res = Optional.empty();
    if ( objects.containsKey(key) ) {
      final Optional<DictNode> k = Optional.ofNullable(objects.get(key));
      res = Optional.of(k.get().string_value);
    }
    return res;
  }
  
  public Optional<Double> getDouble( String key ) {
    Optional<Double> res = Optional.empty();
    if ( objects.containsKey(key) ) {
      final Optional<DictNode> k = Optional.ofNullable(objects.get(key));
      res = Optional.of(k.get().double_value);
    }
    return res;
  }
  
  public Optional<Integer> getInt( String key ) {
    Optional<Integer> res = Optional.empty();
    if ( objects.containsKey(key) ) {
      final Optional<DictNode> k = Optional.ofNullable(objects.get(key));
      res = Optional.of(k.get().int_value);
    }
    return res;
  }
  
  public Optional<Boolean> getBoolean( String key ) {
    Optional<Boolean> res = Optional.empty();
    if ( objects.containsKey(key) ) {
      final Optional<DictNode> k = Optional.ofNullable(objects.get(key));
      res = Optional.of(k.get().boolean_value);
    }
    return res;
  }
  
  public Optional<DictNode> getArray( String key ) {
    Optional<DictNode> res = Optional.empty();
    if ( objects.containsKey(key) ) {
      final Optional<DictNode> obj = Optional.ofNullable(objects.get(key));
      if ( obj.get().is_property ) {
        res = obj.get().object_value;
      }
    }
    return Optional.ofNullable((res.isPresent() ? (DictNode)res.get() : null ) );
  }
  
  public Optional<DictNode> getArrayAt( int index ) {
    Optional<DictNode> res = Optional.empty();
    if ( index < (children.size()) ) {
      res = Optional.of(children.get(index));
    }
    return Optional.ofNullable((res.isPresent() ? (DictNode)res.get() : null ) );
  }
  
  public Optional<DictNode> getObject( String key ) {
    Optional<DictNode> res = Optional.empty();
    if ( objects.containsKey(key) ) {
      final Optional<DictNode> obj = Optional.ofNullable(objects.get(key));
      if ( obj.get().is_property ) {
        res = obj.get().object_value;
      }
    }
    return Optional.ofNullable((res.isPresent() ? (DictNode)res.get() : null ) );
  }
  
  public Optional<DictNode> getObjectAt( int index ) {
    Optional<DictNode> res = Optional.empty();
    if ( index < (children.size()) ) {
      res = Optional.of(children.get(index));
    }
    return Optional.ofNullable((res.isPresent() ? (DictNode)res.get() : null ) );
  }
  
  public String stringify() {
    if ( is_property ) {
      if ( value_type == 7 ) {
        return (("\"" + vref) + "\"") + ":null";
      }
      if ( value_type == 4 ) {
        if ( boolean_value ) {
          return ((("\"" + vref) + "\"") + ":") + "true";
        } else {
          return ((("\"" + vref) + "\"") + ":") + "false";
        }
      }
      if ( value_type == 1 ) {
        return ((("\"" + vref) + "\"") + ":") + double_value;
      }
      if ( value_type == 2 ) {
        return ((("\"" + vref) + "\"") + ":") + int_value;
      }
      if ( value_type == 3 ) {
        return ((((("\"" + vref) + "\"") + ":") + "\"") + this.EncodeString(string_value)) + "\"";
      }
    } else {
      if ( value_type == 7 ) {
        return "null";
      }
      if ( value_type == 1 ) {
        return "" + double_value;
      }
      if ( value_type == 2 ) {
        return "" + int_value;
      }
      if ( value_type == 3 ) {
        return ("\"" + this.EncodeString(string_value)) + "\"";
      }
      if ( value_type == 4 ) {
        if ( boolean_value ) {
          return "true";
        } else {
          return "false";
        }
      }
    }
    if ( value_type == 5 ) {
      String str = "";
      if ( is_property ) {
        str = (("\"" + vref) + "\"") + ":[";
      } else {
        str = "[";
      }
      for ( int i = 0; i < children.size(); i++) {
        DictNode item = children.get(i);
        if ( i > 0 ) {
          str = str + ",";
        }
        str = str + item.stringify();
      }
      str = str + "]";
      return str;
    }
    if ( value_type == 6 ) {
      String str_1 = "";
      if ( is_property ) {
        return ((("\"" + vref) + "\"") + ":") + object_value.get().stringify();
      } else {
        str_1 = "{";
        for ( int i_1 = 0; i_1 < keys.size(); i_1++) {
          String key = keys.get(i_1);
          if ( i_1 > 0 ) {
            str_1 = str_1 + ",";
          }
          final Optional<DictNode> item_1 = Optional.ofNullable(objects.get(key));
          str_1 = str_1 + item_1.get().stringify();
        }
        str_1 = str_1 + "}";
        return str_1;
      }
    }
    return "";
  }
}
